src/ast/expr_ext.rs

```diff
@@ -0,0 +1,45 @@
+use crate::{
+    ast::{self, support::*, AstNode, AstToken},
+    kinds::SyntaxKind::*,
+    match_ast, SyntaxElement,
+};
+
+use super::support::first;
+
+// this is a separate type because it mixes tokens and nodes
+// for example, a Str is a node because it is complex to parse but an Integer is a node.
+// This means that we have to write it out manually instead of using the macro to create the type for us.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub enum LiteralKind {
+    Str(ast::Str),
+    Float(ast::Float),
+    Integer(ast::Integer),
+    Path(ast::Path),
+    Uri(ast::Uri),
+}
+
+impl ast::Literal {
+    pub fn kind(&self) -> LiteralKind {
+        if let Some(it) = first(self) {
+            return LiteralKind::Str(it);
+        }
+
+        if let Some(it) = token(self) {
+            return LiteralKind::Float(it);
+        }
+
+        if let Some(it) = token(self) {
+            return LiteralKind::Integer(it);
+        }
+
+        if let Some(it) = token(self) {
+            return LiteralKind::Path(it);
+        }
+
+        if let Some(it) = token(self) {
+            return LiteralKind::Uri(it);
+        }
```

Comment:
this code doesn't make sense yet.

Reply:
Type inference is one of the best things about rust. It is common to just `.into` without using annotations. What is confusing to you?

Reply:
Type inference is one of the best things about rust. It is common to just `.into` without using annotations. What is confusing to you?

Reply:
The problem is that this looks like weird repetition, and it is necessary to know that `token` is generic over the return parameter to be able to understand what's going on. Transforming it into an `else if` chain would also make the purpose a bit clearer, maybe also add a comment.

e.g.
```rust
        if let Some(it) = token(self) {
            return LiteralKind::Float(it);
        } else if let Some(it) = token(self) {
            return LiteralKind::Integer(it);
        } else if let Some(it) = token(self) {
            return LiteralKind::Path(it);
        } else if let Some(it) = token(self) {
            return LiteralKind::Uri(it);
        }
```

could we instead match on the token type directly?

Reply:
I don't see how the else if makes it better. If it is confusing, you can just hover over the type. `rust-analyzer` also shows really nice inlay type annotations, which clearly show the different types.

Reply:
My complaint is that most people who try to look into the code will stumble over this, especially when unfamiliar, at least it makes the code harder to review.

Reply:
I will accept a comment